mlir/unittests/Example/Example.cpp

```diff
@@ -0,0 +1,49 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Quantum/IR/QuantumOps.h"
+#include "mlir/AsmParser/AsmParser.h"
+#include "mlir/AsmParser/AsmParserState.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Verifier.h"
+#include "mlir/Parser/Parser.h"
+#include "llvm/Support/SourceMgr.h"
+
+#include "gtest/gtest.h"
+
+using namespace mlir;
+
+namespace {
+TEST(MLIRParser, ParseQuantumIR)
+{
+    std::string moduleStr = R"mlir(
+func.func @test_alloc_dealloc_no_fold() -> !quantum.bit {
+  %r = quantum.alloc(3) : !quantum.reg
+  %q = quantum.extract %r[0] : !quantum.reg -> !quantum.bit
+  quantum.dealloc %r : !quantum.reg
+  return %q : !quantum.bit
+}
+  )mlir";
+
+    DialectRegistry registry;
+    registry.insert<func::FuncDialect, catalyst::quantum::QuantumDialect>();
+    MLIRContext context(registry);
+
+    ParserConfig config(&context, /*verifyAfterParse=*/false);
+
+    OwningOpRef<ModuleOp> module = parseSourceString<ModuleOp>(moduleStr, config);
+    ASSERT_TRUE(module);
```

Comment:
Just to make sure I understand the usage:

I think now this `module` is a (or can be cast to a) `ModuleOp`, which allows us to perform transformations/analysis on it?

In which case, we can collect analysis result objects or transformation output IR as strings, and `ASSERT_TRUE(<some callback checking various testing goals>)` on them?

Is there anyway to pipe the transformed IR into filecheck, instead of writing a bunch of string parsing/regexes?

Reply:
`module` should be a `ModuleOp`. But I believe that in the LLVM project, tests that go in the unittests are not intended to be similar to `lit` tests. If that were the case, just write a lit test. Here, there are other things that can be tested. Like for example, testing `SymbolTable::lookup` or testing other data structures that help during the transformation of the IR.

Reply:
Makes sense, I guess we can just check properties of various objects directly in the code, e.g. `ASSERT_TRUE(transformed_funcop.getSymName() == "my_new_name")`, etc.

The reason I mention is because when we extend our dialects, e.g. #1180 , to test the newly added dialect API we need to somehow use them. Creating a specific tester pass just to use that API feels wrong. I think with this unit test infrastructure the aforementioned approach can be used.

Reply:
I wouldn't personally test the builders. But yes, things of that sort may be tested in these unittests.